File: Rust/src/day01.rs
use crate::advent_helper::advent_helper::parse_line_groups;
use crate::advent_helper::advent_helper::print_soln;

fn calorie_count(elf: &String, sep: &str) -> i64
{
	return elf.split(sep).map(|x| x.parse::<i64>().unwrap()).sum::<i64>();
}

pub fn run()
{
	let elfs = parse_line_groups("../input/input01.txt", "|");
	let mut cals: Vec<_> = elfs.iter().map(|x| calorie_count(x, "|")).collect();
	cals.sort_by(|a, b| b.cmp(a));

	print_soln(1, cals[0], cals.iter().take(3).sum::<i64>());
	return;
}

File: C++/Day01.cpp
#include "AH.h"

namespace Day01
{

	int calorieCount(const std::string cals, const char sep) {
		int total = 0;
		const auto cs = AH::Split(cals, sep);
		for (auto & c : cs) {
			total += std::stoi(c);
		}

		return total;
	}

	int Run(const std::string& filename)
	{
		const auto inputLines = AH::ReadTextFile(filename);
		const auto elfs = AH::ParseLineGroups(inputLines, '|');

		std::vector<int> cals;
		cals.reserve(elfs.size());

		for (auto e : elfs) {
			cals.push_back(calorieCount(e, '|'));
		}

		std::sort(cals.begin(), cals.end(), std::greater{});
		
		AH::PrintSoln(1, cals[0], cals[0] + cals[1] + cals[2]);

		return 0;
	}

}
